src/lib/NonlinearDutchDecayLib.sol

```diff
@@ -0,0 +1,126 @@
+// SPDX-License-Identifier: GPL-2.0-or-later
+pragma solidity ^0.8.0;
+
+import {OutputToken, InputToken} from "../base/ReactorStructs.sol";
+import {V3DutchOutput, V3DutchInput, NonlinearDutchDecay} from "../lib/V3DutchOrderLib.sol";
+import {FixedPointMathLib} from "solmate/src/utils/FixedPointMathLib.sol";
+import {MathExt} from "./MathExt.sol";
+import {Uint16ArrayLibrary, Uint16Array, fromUnderlying} from "../types/Uint16Array.sol";
+import {DutchDecayLib} from "./DutchDecayLib.sol";
+
+/// @notice helpers for handling non-linear dutch order objects
+library NonlinearDutchDecayLib {
+    using FixedPointMathLib for uint256;
+    using MathExt for uint256;
+    using Uint16ArrayLibrary for Uint16Array;
+
+    /// @notice thrown when the decay curve is invalid
+    error InvalidDecayCurve();
+
+    /// @notice locates the current position on the curve and calculates the decay
+    /// @param curve The curve to search
+    /// @param startAmount The absolute start amount
+    /// @param decayStartBlock The absolute start block of the decay
+    /// @dev Expects the relativeBlocks in curve to be strictly increasing
+    function decay(
+        NonlinearDutchDecay memory curve,
+        uint256 startAmount,
+        uint256 decayStartBlock,
+        uint256 minAmount,
+        uint256 maxAmount
+    ) internal view returns (uint256 decayedAmount) {
+        // mismatch of relativeAmounts and relativeBlocks
+        if (curve.relativeAmounts.length > 16) {
+            revert InvalidDecayCurve();
+        }
+
+        // handle current block before decay or no decay
+        if (decayStartBlock >= block.number || curve.relativeAmounts.length == 0) {
+            return startAmount.bound(minAmount, maxAmount);
+        }
+        Uint16Array relativeBlocks = fromUnderlying(curve.relativeBlocks);
+        uint16 blockDelta = uint16(block.number - decayStartBlock);
+        int256 curveDelta;
+        // Special case for when we need to use the decayStartBlock (0)
```

Comment:
I think this comment could use a little more color, taking me some effort to figure out what's happening here

So decayStartBlock is already past, but the first block delta hasn't been hit yet? So you're anchoring the left end on block 0? I'm not sure I see the reason even for linear decay there, since it's just going to be sooo close to startAmount + relativeAmounts[0] right?

Reply:
edit: ohh it's _blockDelta_ of 0, not block 0 right?

Reply:
this is one case where maybe named parameters for linearDecay would help

Reply:
Possible to just contain this case within locateCurvePosition?

Reply:
Just refactored it.. lmk what you thnk!